### DISCUSSION TITLE
Importância do ToList - Entity Framework

### DISCUSSION THREAD

**Discussion by pedro-mileipp** [balta-io/forum#257] [Tira Dúvidas] [answered]
- Boa noite, prezado!
- Apesar de você ter explicado no vídeo ainda não tive plena compreensão. Gostaria de saber qual a importância do ToList() no Entity Framework, tendo em vista que com ou não o resultado é o mesmo, assim como por que é importante usar o Where antes do ToList().

```csharp
/* TO LIST */
var tags = context.Tags.ToList(); // executando a Query com o ToList()
foreach(var tag in tags){
    Console.WriteLine("Name: {0} - Slug: {1}",tag.Name, tag.Slug);
}
 Console.WriteLine();

var tags2 = context.Tags.Where(x => x.Name.Contains(".NET")).ToList();
foreach(var tag in tags2){
    Console.WriteLine("Name: {0} - Slug: {1}",tag.Name, tag.Slug);
}
```

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @pedro-mileipp, como vai?

Quando você chama o método `ToList()` o EF materializa (Executa) a query. Isto quer dizer que daqui pra frente, o resultado já está completo, os dados já foram obtidos do banco.

Agora imagina que você está construindo uma biblioteca que vai ser utilizada por N sistemas na sua empresa, ou mesmo que será um pacote público, utilizado por milhões de projetos.

Por um descuido, você esqueceu de retornar uma lista e acabou retornando o DbSet ou um IQueryable. Parece algo sutil, mas nestes casos, sua query ainda não foi materializada, não foi executada no banco.

Se alguém for consumir este pacote, e a partir da sua lista por exemplo, fizer um `Include` de outra tabela com milhões de registros, ele trava o banco. Além disso, dado que a query ainda não foi executada, ela pode ser modificada, incluindo valores que você não quer que sejam acessados ou mesmo acrescentando itens que não fazem sentido para performance e otimização da mesma.

Ficou mais claro agora?
Qualquer dúvida, posta aí.
